File: Surrounded Regions.py
class Solution:
	# @param board, a 2D array
	# Capture all regions by modifying the input board in-place.
	# Do not return any value.
	def solve(self, board):
		M = len(board)
		if (M == 0):
			return
		N = len(board[0])
		dirty = True
		while (dirty):
			dirty = False
			for i in xrange(M):
				for j in xrange(N):
					if (board[i][j] == 'O' and 
						(i == 0 or i == M - 1 or j == 0 or j == N - 1
							or i > 0 and board[i-1][j] == 'v' or j > 0 and board[i][j-1] == 'v'
							or i < M - 1 and board[i+1][j] == 'v' or j < N - 1 and board[i][j+1] == 'v')
						):
						board[i][j] = 'v'
						dirty = True
		for i in xrange(M):
			for j in xrange(N):
				if (board[i][j] == 'O'):
					board[i][j] = 'X'
				elif (board[i][j] == 'v'):
					board[i][j] = 'O'
		
							
					

File: Surrounded Regions.java
public class Solution {
    public void solve(char[][] board) {
        int M = board.length;
        if (M == 0)
            return;
        int N = board[0].length;
        boolean dirty = true;
        while (dirty) {
            dirty = false;
            for (int i = 0; i < M; i++) {
                for (int j = 0; j < N; j++) {
                    if (board[i][j] == 'O' && (i==0 || j==0 || i==M-1 || j==N-1
                        || board[i-1][j]=='R' || board[i+1][j]=='R' || board[i][j-1]=='R' || board[i][j+1]=='R')) {
                        board[i][j] = 'R';
                        dirty = true;
                    }
                }
            }
        }
        for (int i = 0; i < M; i++) {
            for (int j = 0; j < N; j++) {
                if (board[i][j] == 'O')
                    board[i][j] = 'X';
                else if (board[i][j] == 'R')
                    board[i][j] = 'O';
            }
        }
    }
}